### PR TITLE
Cache manager refactor

### PR DESCRIPTION
**Pull request by c-w**
@cpeel: I did a bit of refactoring on your metadata cache manager code in order to gain an understanding of it. I'd like to get some feedback on whether I understood everything correctly before I merge.

The gist of this refactor:
- Create sqlite/bsddb subclasses of MetadataCache to avoid store-type checking in methods
- Create testing MetadataCache to remove testing-only code in the api.
- Fall back to sqlite cache when the user doesn't have bsddb installed.

### CODE REVIEW

File: gutenberg/acquire/metadata.py [hunk 1]
```diff
@@ -162,55 +138,104 @@ def _download_metadata_archive(self):
...
+    def _local_storage_path(self):
+        return self.cache_uri[len(self._CACHE_URI_PREFIX):]
 
-_METADATA_CACHE_MANAGER = MetadataCacheManager(
-    store='Sleepycat',
-    cache_uri=local_path(os.path.join('metadata', 'metadata.db')))
 
+_METADATA_CACHE = None
 
-def set_metadata_cache_manager(cache_manager):
```
Comment (cpeel) [inline gutenberg/acquire/metadata.py]: The set_metadata_cache() function is *not* just for testing, but rather an integral part of my branch. Without a way to explicitly set a cache, you can't control where the cache is stored on the filesystem, or in the new code, control which of the two backend types you want to use. This level of control is vital when using this as part of a larger service.

For an example of what I'm trying to achieve with this module, see:
https://github.com/cpeel/openlibrary/blob/095df4b3be522eef218813a0106ee2ce7934f120/openlibrary/plugins/gutenberg/pgebook.py
and
https://github.com/cpeel/openlibrary/blob/73533a3e011074e2ffbfe416f241a6aa14fb5b6e/scripts/dev-instance/populate_gutenberg_cache.py

**Comment by cpeel** [comment]
Overall I like your refactor -- very clean. My one **big** concern is the removal of `set_metadata_cache[_manager]()` as that function enables one of the core requirements of what I was trying to achieve -- the ability to control where the cache ends up on the filesystem and what type of cache it is.

File: gutenberg/_domain_model/exceptions.py [hunk 2]
```diff
@@ -23,9 +23,5 @@ class CacheAlreadyExists(Error):
     pass
 
 
-class CacheNotRemovable(Error):
```
Comment (cpeel) [inline gutenberg/_domain_model/exceptions.py]: Built-in python errors (almost) all end in Exception, Error, or Warning (https://docs.python.org/2/library/exceptions.html). See also PEP8: https://www.python.org/dev/peps/pep-0008/#exception-names

Comment (c-w) [inline gutenberg/_domain_model/exceptions.py]: Fixed.

File: gutenberg/acquire/metadata.py (hunk 1 above)
Comment (c-w) [inline gutenberg/acquire/metadata.py]: Fixed in 8df3e24.

**Comment by c-w** [comment]
- Added exception suffixes in 8921ee1.
- Froze dependency on sqlalchemy-rdflib in d887023.
- Brought back set_metadata_cache in 8df3e24.

**Comment by cpeel** [comment]
Ship it! :)

**Comment by c-w** [comment]
```
           (`-,-,
           ('(_,( )
            _   `_'
         __|_|__|_|_
       _|___________|__
      |o o o o o o o o/   
     ~'`~'`~'`~'`~'`~'`~
```

**Pull request merged**